elytron/WFCORE-4360-CredentialStore_Expression_Resolution.adoc

```diff
@@ -0,0 +1,105 @@
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:jkasik@redhat.com[Jan Kasik]
+
+=== Testing By
+
+[ ] Engineering
+[x] QE
+
+=== Affected Projects or Components
+
+Utility classes may be required within the WildFly Elytron project and will be developed as the need is identified, however the primary integration will be performed within the WildFly Core project as this enhancment is predominantly in relation to management model expression resolution.
+
+Additionally all documentation lives within the WildFly project so updates will be applied there accordingly.
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+The general intent for this enhancement is to develop some form of configurable expression resolver.
+
+The expression resolver will reference a specific alias within a specific credential store, this credential will be used to decrypt the expression which was previously encrypted using this credential.  The expression resolver will also contain configuration such as the algorithm to use. 
+
+The expression resolver should support both generated keys and password based encryption.
+
+Using the credential store it will be possible to either add or generate suitable keys to be used for the encryption / decryption of values.  This will need to include: -
+
+ * The standalone elytron tool.
+ * The management operations exposed on the credential store.
+
+Where generated keys are used the tools and management operations should support exporting and importing the keys to allow them to be 'moved' between credential stores.  Where possible this should be achieved using specification defined formats.
+
+Functionality should be added to allow an administrator to encrypt a clear text String so that it can be stored as an expression, this will be added to: -
+
+ * The standalone elytron tool.
+ * New management operations, this may however be against the expression resolver not the credential store. 
+
+The domain mode relationships will need to considered further, where we used the PicketBox vault we defined the vault within the host.xml and used it's values in the domain.xml.  As the credential store can both be defined in the host.xml and the domain.xml it would make sense to make use of the one stored within that descriptor, however this could be impacted based on how the configurable expression resolver is defined.
+
+=== Nice-to-Have Requirements
+
+Support for multiple expression resolvers being defined, allowing the correct one to be identified for a specific expression value.  It may not be possible to implement this in the initial enhancement however this should still be considered to allow it's potential addition at a later point.
+
+It would be nice to cross reference subsystem managed security providers for the expression resolver, however this component needs to be usable at the start of `Stage.RUNTIME` so there will be a limit as to how many subsystem managed resources can be depended upon.
```

Comment:
What would be benefits of this requirement?

Reply:
The expression resolver is likely to be using APIs such as https://docs.oracle.com/javase/8/docs/api/javax/crypto/Cipher.html - the implementations of the algorithms coming from the installed java.security.Provider instances - being able to cross reference registered providers would allow for alternative implementations.

One example could be the administrator prefers to use the BouncyCastle provider for this functionality.

Another possibility which we would not directly cover in this RFE is both the Key and the encryption could be handled by a HSM - this integration again is generally configured via providers.

Having said all that custom providers can still be added to the java.security.policy file so there are different ways to achieve the same although they would not be manageable in the same way.